### ISSUE TITLE
Crash in _createLinkedType on 2.1.0-dev.9.1

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#35043]
Analyzer Feedback from IntelliJ

# Version information

- `IDEA IU-182.4892.20`
- `2.1.0-edge.bf26f760b1bb3d5fea6bda110f6a17b590364120`
- `IU-182.4892.20, JRE 1.8.0_152-release-1248-b19x64 JetBrains s.r.o, OS Mac OS X(x86_64) v10.13.6 unknown, screens 1920x1200  Retina`

# Exception

```
Dart analysis server, SDK version 2.1.0-edge.bf26f760b1bb3d5fea6bda110f6a17b590364120, server version 1.20.5, error: Analysis failed: /Users/kevmoo/source/github/graf/app/lib/graph_widget.dart context: exception_20181102_210718_313
Bad state: The type parameter Key (in package:graph/src/directed_graph.dart;package:graph/src/directed_graph.dart;DirectedGraph;Key) is out of scope.
#0      _createLinkedType (package:analyzer/src/summary/link.dart:207:7)
#1      _storeTypeArguments (package:analyzer/src/summary/link.dart:395:31)
#2      _createLinkedType (package:analyzer/src/summary/link.dart:189:5)
#3      CompilationUnitElementInBuildUnit._storeLinkedType (package:analyzer/src/summary/link.dart:1665:31)
#4      FieldElementForLink_ClassField.link (package:analyzer/src/summary/link.dart:2582:23)
#5      ClassElementForLink_Class.link (package:analyzer/src/summary/link.dart:910:20)
#6      CompilationUnitElementInBuildUnit.link (package:analyzer/src/summary/link.dart:1627:20)
#7      LibraryElementInBuildUnit.link (package:analyzer/src/summary/link.dart:3780:12)
#8      LibraryCycleNode.link (package:analyzer/src/summary/link.dart:3486:15)
#9      LibraryCycleDependencyWalker.evaluate (package:analyzer/src/summary/link.dart:3426:7)
#10     DependencyWalker.walk.strongConnect (package:analyzer/src/summary/link.dart:2199:13)
#11     DependencyWalker.walk.strongConnect (package:analyzer/src/summary/link.dart:2165:24)
#12     DependencyWalker.walk (package:analyzer/src/summary/link.dart:2218:18)
#13     LibraryCycleForLink.ensureLinked (package:analyzer/src/summary/link.dart:3459:42)
#14     Linker.link (package:analyzer/src/summary/link.dart:3982:35)
#15     _relink (package:analyzer/src/summary/link.dart:302:57)
#16     link (package:analyzer/src/summary/link.dart:120:3)
#17     new LibraryContext.forSingleLibrary.<anonymous closure>.<anonymous closure> (package:analyzer/src/dart/analysis/library_context.dart:118:27)
#18     PerformanceLog.run (package:analyzer/src/dart/analysis/performance_logger.dart:34:15)
#19     new LibraryContext.forSingleLibrary.<anonymous closure> (package:analyzer/src/dart/analysis/library_context.dart:117:14)
#20     PerformanceLog.run (package:analyzer/src/dart/analysis/performance_logger.dart:34:15)
#21     new LibraryContext.forSingleLibrary (package:analyzer/src/dart/analysis/library_context.dart:56:19)
#22     AnalysisDriver._createLibraryContext (package:analyzer/src/dart/analysis/driver.dart:1374:16)
<asynchronous suspension>
#23     AnalysisDriver._computeAnalysisResult.<anonymous closure> (package:analyzer/src/dart/analysis/driver.dart:1238:34)
<asynchronous suspension>
#24     PerformanceLog.runAsync (package:analyzer/src/dart/analysis/performance_logger.dart:52:21)
<asynchronous suspension>
#25     AnalysisDriver._computeAnalysisResult (package:analyzer/src/dart/analysis/driver.dart:1224:20)
<asynchronous suspension>
#26     AnalysisDriver.performWork (package:analyzer/src/dart/analysis/driver.dart:1050:39)
<asynchronous suspension>
#27     AnalysisDriverScheduler._run (package:analyzer/src/dart/analysis/driver.dart:1891:24)
<asynchronous suspension>
#28     AnalysisDriverScheduler.start (package:analyzer/src/dart/analysis/driver.dart:1832:5)
#29     new AnalysisServer (package:analysis_server/src/analysis_server.dart:340:29)
#30     SocketServer.createAnalysisServer (package:analysis_server/src/socket_server.dart:79:26)
#31     StdioAnalysisServer.serveStdio (package:analysis_server/src/server/stdio_server.dart:37:18)
#32     Driver.start.<anonymous closure> (package:analysis_server/src/server/driver.dart:505:21)
#33     _rootRun (dart:async/zone.dart:1124:13)
#34     _CustomZone.run (dart:async/zone.dart:1021:19)
#35     _runZoned (dart:async/zone.dart:1516:10)
#36     runZoned (dart:async/zone.dart:1463:12)
#37     Driver._captureExceptions (package:analysis_server/src/server/driver.dart:551:12)
#38     Driver.start (package:analysis_server/src/server/driver.dart:503:7)
#39     main (file:///b/build/slave/Mac_Engine/build/src/third_party/dart/pkg/analysis_server/bin/server.dart:12:11)
#40     _AsyncAwaitCompleter.start (dart:async/runtime/libasync_patch.dart:49:6)
#41     main (file:///b/build/slave/Mac_Engine/build/src/third_party/dart/pkg/analysis_server/bin/server.dart:10:10)
#42     _startIsolate.<anonymous closure> (dart:isolate/runtime/libisolate_patch.dart:287:32)
#43     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:171:12)

```

```
1541218026973 => {"id":"24","method":"edit.getAssists","params":{"file":"/Users/kevmoo/source/github/graf/app/lib/graph_widget.dart","offset":600,"length":0},"clientRequestTime":1541218026973}
1541218027062 => {"id":"25","method":"edit.getAssists","params":{"file":"/Users/kevmoo/source/github/graf/app/lib/graph_widget.dart","offset":600,"length":0},"clientRequestTime":1541218027062}
1541218027063 => {"id":"26","method":"edit.getAssists","params":{"file":"/Users/kevmoo/source/github/graf/app/lib/graph_widget.dart","offset":600,"length":0},"clientRequestTime":1541218027063}
1541218027177 => {"id":"27","method":"analysis.setPriorityFiles","params":{"files":["/Users/kevmoo/source/github/graf/app/lib/node.dart"]},"clientRequestTime":1541218027177}
1541218027177 => {"id":"28","method":"analysis.setSubscriptions","params":{"subscriptions":{"OVERRIDES":["/Users/kevmoo/source/github/graf/app/lib/node.dart"],"CLOSING_LABELS":["/Users/kevmoo/source/g
1541218027675 <= {"id":"24","result":{"assists":[]}}
1541218027675 <= {"id":"25","result":{"assists":[]}}
1541218027675 <= {"id":"26","result":{"assists":[]}}
1541218027676 <= {"id":"27"}
1541218027677 <= {"id":"28"}
1541218027781 <= {"event":"server.status","params":{"analysis":{"isAnalyzing":false}}}
1541218035287 => {"id":"29","method":"server.shutdown","clientRequestTime":1541218035287}
1541218035291 <= {"id":"29"}
1541218035356 started analysis server:
  /Users/kevmoo/source/github/flutter/bin/cache/dart-sdk/bin/dart /Users/kevmoo/source/github/flutter/bin/cache/dart-sdk/bin/snapshots/analysis_server.dart.snapshot --client-id=IntelliJ-IDEA --client-
1541218035718 <= {"event":"server.connected","params":{"version":"1.20.5","pid":54895,"sessionId":""}}
1541218035738 <= {"id":"0","result":{"version":"1.20.5"}}
1541218035739 => {"id":"1","method":"server.setSubscriptions","params":{"subscriptions":["STATUS"]},"clientRequestTime":1541218035369}
1541218035739 => {"id":"2","method":"analysis.updateOptions","params":{"options":{"enableAsync":true,"enableDeferredLoading":true,"enableEnums":true,"enableNullAwareOperators":true,"enableSuperMixins"
1541218035739 => {"id":"3","method":"analysis.setAnalysisRoots","params":{"included":["/Users/kevmoo/source/github/graf/app"],"excluded":["/Users/kevmoo/source/github/graf/app/.idea","/Users/kevmoo/so
1541218035739 => {"id":"4","method":"analysis.setPriorityFiles","params":{"files":["/Users/kevmoo/source/github/graf/app/lib/node.dart"]},"clientRequestTime":1541218035374}
1541218035739 => {"id":"5","method":"analysis.setSubscriptions","params":{"subscriptions":{"OVERRIDES":["/Users/kevmoo/source/github/graf/app/lib/node.dart"],"CLOSING_LABELS":["/Users/kevmoo/source/gi
1541218035742 <= {"id":"1"}
1541218035745 <= {"id":"2"}
1541218035895 <= {"event":"analysis.errors","params":{"file":"/Users/kevmoo/source/github/graf/app/pubspec.yaml","errors":[]}}
1541218035963 <= {"id":"3"}
1541218035966 <= {"id":"4"}
1541218035976 <= {"id":"5"}
1541218035982 <= {"event":"server.status","params":{"analysis":{"isAnalyzing":true}}}
1541218038346 <= {"event":"server.error","params":{"isFatal":false,"message":"Analysis failed: /Users/kevmoo/source/github/graf/app/lib/graph_widget.dart context: exception_20181102_210718_313","stack
```

**Comment by kevmoo** [comment]
Seems to happen when under-providing type arguments. In this case, I provided 2 when the class expected 3

**Comment by mit-mit** [comment]
cc @stereotype441 @bwilkerson

**Comment by stereotype441** [comment]
@kevmoo have you found a repro using the command-line analyzer?  If so that would be really helpful in tracking the bug down.

**Comment by kevmoo** [comment]
Yup! Minimal repo

```dart
final foo = Map<int>();
```

```
Unhandled exception:
Bad state: The type parameter K (in dart:core;dart:core/map.dart;Map;K) is out of scope.
#0      _createLinkedType (package:analyzer/src/summary/link.dart:207:7)
#1      _storeTypeArguments (package:analyzer/src/summary/link.dart:395:31)
#2      _createLinkedType (package:analyzer/src/summary/link.dart:189:5)
#3      CompilationUnitElementInBuildUnit._storeLinkedType (package:analyzer/src/summary/link.dart:1665:31)
#4      TopLevelVariableElementForLink.link (package:analyzer/src/summary/link.dart:5101:25)
#5      CompilationUnitElementInBuildUnit.link (package:analyzer/src/summary/link.dart:1624:16)
#6      LibraryElementInBuildUnit.link (package:analyzer/src/summary/link.dart:3780:12)
#7      LibraryCycleNode.link (package:analyzer/src/summary/link.dart:3486:15)
#8      LibraryCycleDependencyWalker.evaluate (package:analyzer/src/summary/link.dart:3426:7)
#9      DependencyWalker.walk.strongConnect (package:analyzer/src/summary/link.dart:2199:13)
#10     DependencyWalker.walk (package:analyzer/src/summary/link.dart:2218:18)
#11     LibraryCycleForLink.ensureLinked (package:analyzer/src/summary/link.dart:3459:42)
#12     Linker.link (package:analyzer/src/summary/link.dart:3982:35)
#13     _relink (package:analyzer/src/summary/link.dart:302:57)
#14     link (package:analyzer/src/summary/link.dart:120:3)
#15     new LibraryContext.forSingleLibrary.<anonymous closure>.<anonymous closure> (package:analyzer/src/dart/analysis/library_context.dart:118:27)
#16     PerformanceLog.run (package:analyzer/src/dart/analysis/performance_logger.dart:34:15)
#17     new LibraryContext.forSingleLibrary.<anonymous closure> (package:analyzer/src/dart/analysis/library_context.dart:117:14)
#18     PerformanceLog.run (package:analyzer/src/dart/analysis/performance_logger.dart:34:15)
#19     new LibraryContext.forSingleLibrary (package:analyzer/src/dart/analysis/library_context.dart:56:19)
#20     AnalysisDriver._createLibraryContext (package:analyzer/src/dart/analysis/driver.dart:1374:16)
<asynchronous suspension>
#21     AnalysisDriver._computeUnitElement (package:analyzer/src/dart/analysis/driver.dart:1316:43)
<asynchronous suspension>
#22     AnalysisDriver.performWork (package:analyzer/src/dart/analysis/driver.dart:971:40)
<asynchronous suspension>
#23     AnalysisDriverScheduler._run (package:analyzer/src/dart/analysis/driver.dart:1891:24)
<asynchronous suspension>
#24     AnalysisDriverScheduler.start (package:analyzer/src/dart/analysis/driver.dart:1832:5)
#25     Driver._createContextAndAnalyze (package:analyzer_cli/src/driver.dart:658:15)
#26     Driver._analyzeAllImpl (package:analyzer_cli/src/driver.dart:328:9)
<asynchronous suspension>
#27     Driver._analyzeAll (package:analyzer_cli/src/driver.dart:258:20)
<asynchronous suspension>
#28     Driver.start (package:analyzer_cli/src/driver.dart:209:38)
<asynchronous suspension>
#29     main (file:///b/build/slave/dart-sdk-mac-dev/build/sdk/pkg/analyzer_cli/bin/analyzer.dart:17:17)
<asynchronous suspension>
#30     _startIsolate.<anonymous closure> (dart:isolate/runtime/libisolate_patch.dart:285:32)
#31     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:171:12)
```

**Comment by stereotype441** [comment]
Wow, I'm able to reproduce this bug going back at least as far as 2.0.0-dev.69.0.  So even though the repro is so small, I don't think this needs to block the 2.1 release--after all, we released 2.0 with the bug and no one appears to have noticed!

I will keep the P1 designation though, since the repro is so embarrassingly simple.

**Comment by kevmoo** [comment]
ack!
